apps/aecore/test/aec_tx_pool_tests.erl

```diff
@@ -96,6 +96,87 @@ tx_pool_test_() ->
+
+               %% PubKey1 creates two transaction but has enough fund only for one
+               %% PubKey2 has enough fund for all it's transactions
+               Txs1 = [Tx1_1, Tx1_2]  = [ a_signed_tx(PubKey1, me, Nonce, 20000, 10) || Nonce <- lists:seq(1, 2) ],
+               Txs2 = [Tx2_1, Tx2_2]  = [ a_signed_tx(PubKey2, me, Nonce, 20000, 10) || Nonce <- lists:seq(1, 2) ],
+               [ ok = aec_tx_pool:push(Tx, tx_created) || Tx <- Txs1++Txs2 ],
+               ?assertMatch([_,_,_,_], aec_tx_pool:peek_db()),
+               ?assertEqual([], aec_tx_pool:peek_visited()),
+
+               %% Micro block candidate contains all txs because
+               %% total gas for them is smaller then aec_governance:block_gas_limit()
+               {ok, CTxs1} = aec_tx_pool:get_candidate(aec_governance:block_gas_limit(), KeyHash),
+               ?assertEqual(lists:sort(Txs1++Txs2), lists:sort(CTxs1)),
+               ?assertEqual([], aec_tx_pool:peek_db()),
+               ?assertMatch([_,_,_,_], aec_tx_pool:peek_visited()),
+
+               {ok, MicroCandidate, _} = aec_block_micro_candidate:create(KeyBlock),
+               {ok, Micro} = aec_keys:sign_micro_block(MicroCandidate),
+               ok = aec_chain_state:insert_block(Micro),
+               {ok, MicroHash} = aec_blocks:hash_internal_representation(Micro),
+               ?assertEqual(MicroHash, aec_chain:top_block_hash()),
+
+               %% Micro block contains only valid transacions
+               %% Tx1_2 is excluded because PubKey1 account has not enough funds
+               ?assert(lists:member(Tx1_1, aec_blocks:txs(Micro))),
+               ?assert(not lists:member(Tx1_2, aec_blocks:txs(Micro))),
+               ?assert(lists:member(Tx2_1, aec_blocks:txs(Micro))),
+               ?assert(lists:member(Tx2_2, aec_blocks:txs(Micro))),
+
+               ?assertEqual([], aec_tx_pool:peek_db()),
+               ?assertMatch([_,_,_,_], aec_tx_pool:peek_visited()),
+
+               aec_tx_pool:top_change(micro, KeyHash, MicroHash),
+
+               %% Invalid Txs1_2 tx is still in the pool
+               ?assertEqual([], aec_tx_pool:peek_db()),
+               ?assertMatch([Tx1_2], aec_tx_pool:peek_visited()),
+
+               %% No new transaction: retry previously invalid tx
+               ?assertEqual([], aec_tx_pool:peek_db()),
+               {ok, CTxs2} = aec_tx_pool:get_candidate(aec_governance:block_gas_limit(), MicroHash),
+               ?assertEqual([Tx1_2], CTxs2),
+
+               %% Some new transacions (new tx gas < gas limit - invalid tx gas): use new + retry previously invalid tx
+               ?assertEqual([], aec_tx_pool:peek_db()),
+               Txs3 = [ a_signed_tx(PubKey2, me, Nonce, 20000, 10) || Nonce <- lists:seq(3, 103) ],
+               [ ok = aec_tx_pool:push(Tx) || Tx <- Txs3 ],
+               TotalGas3 = lists:sum([ aetx:gas(aetx_sign:tx(T)) || T <- [ Tx1_2 | Txs3 ] ]),
```

Comment:
This will conflict https://github.com/aeternity/epoch/pull/1796/files#diff-b7fc65d08fbfee442d57ae86b1b2576cR15

I suggest we merge this after the other one - as this is test only and shall be easier to rebase.